client/src/controllers/EnableDirtyFormCheckController.ts

```diff
@@ -0,0 +1,190 @@
+              this.commentApp.store.getState(),
+            );
+            if (newIsCommentsDirty !== this.isCommentsDirty) {
+              this.isCommentsDirty = newIsCommentsDirty;
+              this.updateCallback(this.isDirty, this.isCommentsDirty);
+            }
+          },
+          this.isCommentsDirty ? 3000 : 300,
+        );
+      });
+    }
+
+    this.updateCallback(this.isDirty, this.isCommentsDirty);
+    this.isDirtyTimeout = -1;
+
+    if (!this.alwaysDirtyValue) {
+      setTimeout(() => {
+        this.initialData = this.getFormData();
+
+        const updateDirtyCheck = () => {
+          clearTimeout(this.isDirtyTimeout);
+          // If the form is dirty, it is relatively unlikely to become clean again, so
+          // run the dirty check on a relatively long timer that we reset on any form update
+          // otherwise, use a short timer both for nicer UX and to ensure widgets
+          // like Draftail have time to serialize their data
+          this.isDirtyTimeout = setTimeout(
+            this.updateIsDirty.bind(this),
+            this.isDirty ? 3000 : 300,
+          );
+        };
+
+        this.element.addEventListener('change', updateDirtyCheck);
+        this.element.addEventListener('keyup', updateDirtyCheck);
+        updateDirtyCheck();
+
+        const isValidInputNode = (node) =>
+          node.nodeType === node.ELEMENT_NODE &&
+          ['INPUT', 'TEXTAREA', 'SELECT'].includes(node.tagName);
+
+        this.observer = new MutationObserver((mutationList) => {
+          const hasMutationWithValidInputNode = mutationList.some(
+            (mutation) =>
+              Array.from(mutation.addedNodes).some(isValidInputNode) ||
+              Array.from(mutation.removedNodes).some(isValidInputNode),
+          );
+
+          if (hasMutationWithValidInputNode) {
+            updateDirtyCheck();
+          }
+        });
+
+        this.observer.observe(this.element, {
+          childList: true,
+          attributes: false,
+          subtree: true,
+        });
+      }, 1000 * 10);
+    }
+  }
+
+  disconnect(): void {
+    this.observer.disconnect();
+  }
+
+  beforeLeaving(event: BeforeUnloadEvent): void {
+    clearTimeout(this.isDirtyTimeout);
+    this.updateIsDirty();
+    const displayConfirmation =
+      !this.formSubmitted && (this.isDirty || this.isCommentsDirty);
+
+    if (displayConfirmation) {
+      event.preventDefault();
+      // eslint-disable-next-line no-param-reassign
+      event.returnValue = this.confirmationMessageValue;
+    }
+  }
+
+  getFormData() {
+    const formMapped = new Map<string, Array<string | File>>();
+    const formData = new FormData(this.element);
+    Array.from(formData.keys())
+      .filter((key) => !key.startsWith('comments-'))
+      .forEach((key) => formMapped.set(key, formData.getAll(key)));
+    return formMapped;
+  }
+
+  updateCallback(formDirty: boolean, commentsDirty: boolean) {
```

Comment:
I think we can simplify this 'reactive' code. 

Stimulus gives us a great way to make a controller reactive to changes in its own values. It's the value changed callbacks. 

As per other notes, if we have two values (comment, form) that represent either a form value or a comment has changed. We can then use the `formValueChanged` and `commentsValueChanged` methods to listen for these changes. 

If we have both these methods working similarly

```js
formValueChanged(current, previous) {
  // .. do the always dirty check as above. 
 if (current !== previous) {
    this.updateMessage();
    return;
  }

}

formValueChanged(current, previous) {
  if (previous=== undefined) return;
  if (current !== previous) {
    this.updateMessage();
    return;
  }
}

updateMessage() {
  this.commentsValue:...
  this.isDirty etc
  // Dispatches the events for the message add / clear
}
```

Now. This creates a problem when updating things close together, the message update method may run too much. 


But we can solve this with making the updateMessage method a deferred method using a debounce wrapper. See the example below. 

https://github.com/wagtail/wagtail/blob/8a6ab175fc1ee2eab53f7223a5bb24e97a2a278f/client/src/controllers/AutosizeController.ts#L20